logitech-scribe: add wait for replug

vcdmp opened the pull request:
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [x] Code fix
- [ ] Feature
- [ ] Documentation

fwupdmgr reporting same firmware, after the upgrade. Added missing FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG 
Moved config setting from quirk to device class

hughsie commented on plugins/logitech-scribe/fu-logitech-scribe-device.c:
```diff
@@ -561,6 +562,9 @@ fu_logitech_scribe_device_write_firmware(FuDevice *device,
 	 * image file pushed. Device validates and uploads new image on inactive partition. Reboots
 	 * wait for RemoveDelay duration
 	 */
+	fu_device_sleep_full(FU_DEVICE(self), 60 * 1000, fu_progress_get_child(progress));
```
I'm confused what this sleep is achieving? Doesn't it work without this? All this is doing is blocking the mainloop so that the uevent is processed slightly later.

vcdmp replied:
Sleep is needed. It doesn't solve the issue without Sleep. By the time code hit Sleep, device is already gone/disconnected/rebooting.

hughsie replied:
Okay, I'm really confused why the sleep would work here -- is the device somehow getting confused when we probe it as soon as it enumerates as a new device? e.g. I'd be interested in the difference between `sudo fwupdtool --plugins logitech_scribe install-blob THE_PAYLOAD -vv --no-timestamp` between the sleep and no-sleep variants. i.e. this smells like a workaround that we need to investigate a bit more.

vcdmp commented:
Here are requested log files. Without sleep, fwupd failed to detect rebooted device. Firmware does gets updated successfully though. Seeing same behavior with daemons (fwupd+fwupdmgr)

[fwupdtool_vv_with_without_fu_device_sleep_full.zip](https://github.com/user-attachments/files/20137584/fwupdtool_vv_with_without_fu_device_sleep_full.zip)

hughsie commented:
I think the device list is confused as you're not actually setting the remove delay anywhere. i.e. if you add something like `fu_device_set_remove_delay(FU_DEVICE(self), FU_DEVICE_REMOVE_DELAY_RE_ENUMERATE)` in `fu_logitech_scribe_device_init` (or maybe a bigger number like 12000ms if the device does take a long time to come back) can you remove the manual delay? It's a bit of a footgun forgetting to set the delay, so I'll add something in the daemon to make some noise if the plugin author forgets.

hughsie commented:
e.g. https://github.com/fwupd/fwupd/pull/8796

hughsie commented:
@vcdmp also; if you've got hardware connected -- could you record an emulation we could use for CI coverage testing? e.g. https://github.com/fwupd/fwupd/blob/main/docs/device-emulation.md

vcdmp commented:
> @vcdmp also; if you've got hardware connected -- could you record an emulation we could use for CI coverage testing? e.g. https://github.com/fwupd/fwupd/blob/main/docs/device-emulation.md

vcdmp commented:
Tried following scenarios, but none helping:
Add fu_device_set_remove_delay(FU_DEVICE(self), 60 * 1000);
Add fu_device_set_remove_delay(FU_DEVICE(self), 2 * 60 * 1000);
Add fu_device_set_remove_delay(FU_DEVICE(self), FU_DEVICE_REMOVE_DELAY_RE_ENUMERATE);
Remove 'RemoveDelay = 120000' line from quirk file

Attached new log for modified code currently in PR:
[fwupdtool_vv.zip](https://github.com/user-attachments/files/20235590/fwupdtool_vv.zip)

hughsie commented:
>  Remove 'RemoveDelay = 120000' line from quirk file

Ohh I missed that, apologies -- any reason you added it there as a quirk entry rather than in fu_logitech_scribe_device_init()? Using the quirk is quite a bit less efficient than hardcoding it; that said: it should still work. Could you recapture the both logs (i.e. with and without the added delay) **without** --no-timestamps please? I want to see the timings of what's happening -- there's something weird going on and I'm not keen on a bandaid.

vcdmp commented:
Leveraging quirk entry seems less intrusive. Removed it.
Removed "--no-timestamps", I was surprised as well, when you originally requested to run with "--no-timestamps"

Attached new log for modified code currently in PR. Only difference is fu_device_sleep_full is commented out to capture the fail usecase. 

Please ignore "percentage should not go down" for now, will revisit once we have consensus here.

[fwupdtool_vv_with_without_fu_device_sleep_full.zip](https://github.com/user-attachments/files/20258126/fwupdtool_vv_with_without_fu_device_sleep_full.zip)

vcdmp commented:
[PR](https://github.com/fwupd/fwupd/pull/8865) didn't helped. 
For troubleshooting purpose, I added debug statement in that loop, but it just executed once, so no pending events.

Attached new logs, with & without sleep.
[fwupdtool_vv_with_without_fu_device_sleep_full.zip](https://github.com/user-attachments/files/20583034/fwupdtool_vv_with_without_fu_device_sleep_full.zip)

vcdmp commented:
> > @vcdmp also; if you've got hardware connected -- could you record an emulation we could use for CI coverage testing? e.g. https://github.com/fwupd/fwupd/blob/main/docs/device-emulation.md

Will do, once we have this fix in place.

hughsie commented:
@vcdmp did you mean to take out the sleep?

vcdmp commented:
We need successful firmware update. If leaving Sleep is OK now, then I am fine too.

hughsie commented:
Ya, the sleep is fine -- we've both tried to find the root issue but working firmware updates is the most important thing.

vcdmp commented:
Sorry, I saw you approved. Let me know, if any issues/concern here.
